source/com.microsoft.tfs.client.common.ui/src/com/microsoft/tfs/client/common/ui/dialogs/connect/OAuth2DeviceFlowCallbackDialog.java

```diff
@@ -0,0 +1,129 @@
+import com.microsoft.tfs.client.common.ui.framework.dialog.BaseDialog;
+import com.microsoft.tfs.client.common.ui.framework.helper.SWTUtil;
+import com.microsoft.tfs.client.common.ui.framework.layout.GridDataBuilder;
+
+public class OAuth2DeviceFlowCallbackDialog extends BaseDialog {
+
+    private final DeviceFlowResponse response;
+    private final FormToolkit toolkit;
+
+    public OAuth2DeviceFlowCallbackDialog(final Shell parentShell, final DeviceFlowResponse response) {
+        super(parentShell);
+
+        this.response = response;
+        this.toolkit = new FormToolkit(parentShell.getDisplay());
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @seecom.microsoft.tfs.client.common.ui.shared.dialog.BaseDialog#
+     * hookAddToDialogArea(org.eclipse.swt.widgets.Composite)
+     */
+    @Override
+    protected void hookAddToDialogArea(final Composite dialogArea) {
+        final GridLayout layout = new GridLayout(3, false);
+
+        layout.marginWidth = getHorizontalMargin();
+        layout.marginHeight = getVerticalMargin();
+        layout.horizontalSpacing = getHorizontalSpacing();
+        layout.verticalSpacing = getVerticalSpacing();
+        dialogArea.setLayout(layout);
+
+        final Label verificationUrlLabel =
+            SWTUtil.createLabel(dialogArea, Messages.getString("OAuth2DeviceFlowCallbackDialog.DeviceLoginUrlText")); //$NON-NLS-1$
+        GridDataBuilder.newInstance().hSpan(layout).hFill().hGrab().applyTo(verificationUrlLabel);
+
+        final String urlLinkText = this.response.getVerificationUri().toString();
+        final Hyperlink verificationUrlLink =
+            this.toolkit.createHyperlink(dialogArea, urlLinkText, SWT.WRAP | SWT.CENTER);
+        GridDataBuilder.newInstance().hSpan(layout).hAlignCenter().applyTo(verificationUrlLink);
+
+        verificationUrlLink.setUnderlined(false);
+        verificationUrlLink.addHyperlinkListener(new HyperlinkAdapter() {
+            @Override
+            public void linkActivated(final HyperlinkEvent e) {
+                try {
+                    PlatformUI.getWorkbench().getBrowserSupport().getExternalBrowser().openURL(new URL(e.getLabel()));
+                } catch (PartInitException e1) {
+                    e1.printStackTrace();
+                } catch (MalformedURLException e1) {
+                    e1.printStackTrace();
+                }
+            }
+        });
+
+        final Label codeNameLabel =
+            SWTUtil.createLabel(dialogArea, Messages.getString("OAuth2DeviceFlowCallbackDialog.DeviceUserCodeText")); //$NON-NLS-1$
+        GridDataBuilder.newInstance().hSpan(layout).hFill().hGrab().applyTo(codeNameLabel);
+
+        final Text deviceCodeText = new Text(dialogArea, SWT.READ_ONLY | SWT.BORDER);
+        GridDataBuilder.newInstance().hSpan(layout).hFill().applyTo(deviceCodeText);
+        deviceCodeText.setText(this.response.getUserCode());
+        deviceCodeText.setFocus();
```

Comment:
Yeah it's selected by default since it has focus.  Just need to "ctrl + c" to copy it.